Allow Collection for 'fields' and 'exclude' of form model helpers (#637)

mthuurne opened the pull request:
## Related issues

Closes #637

There are several functions and classes in `django.forms.models` that take a `fields` or `exclude` argument. Previously, `Sequence` was used to annotate these, but the code of Django (I checked version 3.2.4) doesn't require `__getitem__()` to be implemented, so requiring `Collection` instead is sufficient.

The practical advantage of requiring `Collection` is that a set, such as the key set of a dictionary, can be passed without first having to convert it to a list or tuple.

---

By the way, the full definition of `_Fields` is:
```py
_Fields = Union[List[Union[Callable, str]], Collection[str], Literal["__all__"]]
```
The `Literal["__all__"]` is for accepting `ALL_FIELDS`, but I don't know what `List[Union[Callable, str]]` is for. I couldn't find any code or documentation that treats elements of `fields` and `exclude` as callable.

mthuurne committed:
Allow Collection for 'fields' and 'exclude' of form model helpers (#637)

There are several functions and classes in `django.forms.models` that
take a `fields` or `exclude` argument. Previously, `Sequence` was used
to annotate these, but the code of Django (I checked version 3.2.4)
doesn't require `__getitem__()` to be implemented, so requiring
`Collection` instead is sufficient.

The practical advantage of requiring `Collection` is that a set, such
as the key set of a dictionary, can be passed without first having to
convert it to a list or tuple.

sobolevn commented on django-stubs/forms/models.pyi:
```diff
@@ -36,7 +37,7 @@ from django.db.models import ForeignKey
 ALL_FIELDS: str
 
-_Fields = Union[List[Union[Callable, str]], Sequence[str], Literal["__all__"]]
+_Fields = Union[List[Union[Callable, str]], Collection[str], Literal["__all__"]]
```
Btw, can we improve `Callable` type here? Right now it is just `Callable[..., Any]`. Maybe it can be more specific?

mthuurne replied:
As I wrote in the initial PR message, I don't see the code using `fields` or `exclude` as a callable anywhere, at least not in Django 3.2.4. Are you certain that's a feature that (still) exists?

sobolevn replied:
Ok, let's remove it then! 👍

mthuurne replied:
I pushed a commit containing its removal.

sobolevn reviewed (approved):
Thanks!

mthuurne committed:
Pin mypy to below version 0.900

mthuurne committed:
Remove Callable for 'fields' and 'exclude' of form model helpers

I cannot find any support for callables for these two arguments in
the code or in the documentation.

sobolevn committed:
Update setup.py
